TypeCobol.LanguageServer/Workspace.cs

```diff
@@ -951,6 +953,203 @@ private void FinalCompilationStepCompleted(object cUnit, ProgramClassEvent progr
+
+        /// <summary>
+        /// Get the Data Layout rows for a Program or a Copy (output = CSV)
+        /// </summary>
+        /// <param name="compilationUnit">Compilation unit resulting from parsing the Program/Copy</param>
+        /// <param name="separator">Separator to be used in the rows</param>
+        /// <returns></returns>
+        public string[] GetDataLayoutAsCSV(CompilationUnit compilationUnit, string separator)
+        {
+            var rows = new List<string>();
+            foreach (var dataDefinition in CollectDataLayoutDefinitions(compilationUnit))
+            {
+                var row = CreateRow(dataDefinition, separator);
+                if (row != null)
+                {
+                    rows.Add(row);
+                }
+            }
+
+            return rows.ToArray<string>();
+
+            static string CreateRow(DataDefinition dataDefinition, string separator)
+            {
+                if (IsIgnored(dataDefinition))
+                {
+                    // Ignore this node
+                    return null;
+                }
+
+                var row = new StringBuilder();
+
+                //TODO manage slack bytes (property is dataDefinition.SlackBytes)
+
+                // Line number (starting at 1)
+                AppendToRow(dataDefinition.CodeElement.GetLineInMainSource() + 1);
+
+                /// Level number
+                AppendToRow(dataDefinition.CodeElement.LevelNumber);
+
+                // Name (preceded by an indent depending on the level)
+                AppendToRow(GetIndent(dataDefinition) + GetName(dataDefinition));
+
+                // Declaration (Picture, Usage, REDEFINES, OCCURS, ...)
+                AppendToRow(GetDeclaration(dataDefinition));
+
+                // Start/End/Length
+                var start = dataDefinition.StartPosition;
+                var length = dataDefinition.PhysicalLength;
+                AppendToRow(start);
+                AppendToRow(GetEnd(start, length));
+                AppendToRow(length);
+
+                return row.ToString();
+
+                static bool IsIgnored(DataDefinition dataDefinition)
+                {
+                    DataDefinitionEntry codeElement = dataDefinition.CodeElement;
+                    return codeElement == null || codeElement.Line < 0;
+                }
+
+                void AppendToRow(object value)
+                {
+                    row.Append(value).Append(separator);
+                }
+
+                static string GetIndent(DataDefinition dataDefinition)
+                {
+                    var result = new StringBuilder();
+                    Node parentData = dataDefinition.Parent;
+                    while (parentData is not DataSection)
+                    {
+                        parentData = parentData.Parent;
+                        result.Append(INDENT);
+                    }
+
+                    return result.ToString();
+                }
```

Comment:
I haven't read the CollectDataDefinitions method but I have the intuition this should be done when collecting nodes instead of walking their parents afterward.

Reply:
To be more precise, the collecting method should return a `List<Tuple<int, DataDefinition>>`

Reply:
Good intuition, it's done